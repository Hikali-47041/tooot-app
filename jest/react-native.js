jest.mock('react-native/Libraries/LayoutAnimation/LayoutAnimation', () => ({
  ...require.requireActual(
    'react-native/Libraries/LayoutAnimation/LayoutAnimation'
  ),
  configureNext: jest.fn()
}))
