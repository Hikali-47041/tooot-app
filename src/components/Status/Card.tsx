import React from 'react'
import { Image, Pressable, StyleSheet, Text, View } from 'react-native'
import { useNavigation } from '@react-navigation/native'

export interface Props {
  card: Mastodon.Card
}

const Card: React.FC<Props> = ({ card }) => {
  const navigation = useNavigation()
  return (
    card && (
      <Pressable
        style={styles.card}
        onPress={() => {
          navigation.navigate('Webview', {
            uri: card.url
          })
        }}
      >
        {card.image && (
          <View style={styles.left}>
            <Image source={{ uri: card.image }} style={styles.image} />
          </View>
        )}
        <View style={styles.right}>
          <Text numberOfLines={1}>{card.title}</Text>
          {card.description ? (
            <Text numberOfLines={2}>{card.description}</Text>
          ) : (
            <></>
          )}
          <Text numberOfLines={1}>{card.url}</Text>
        </View>
      </Pressable>
    )
  )
}

const styles = StyleSheet.create({
  card: {
    flex: 1,
    flexDirection: 'row',
    height: 70,
    marginTop: 12
  },
  left: {
    width: 70
  },
  image: {
    width: '100%',
    height: '100%'
  },
  right: {
    flex: 1
  }
})

export default Card
