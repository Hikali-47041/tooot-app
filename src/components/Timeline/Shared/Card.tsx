import analytics from '@components/analytics'
import GracefullyImage from '@components/GracefullyImage'
import openLink from '@components/openLink'
import { useNavigation } from '@react-navigation/native'
import { StyleConstants } from '@utils/styles/constants'
import { useTheme } from '@utils/styles/ThemeManager'
import React from 'react'
import { Pressable, StyleSheet, Text, View } from 'react-native'

export interface Props {
  card: Pick<
    Mastodon.Card,
    'url' | 'image' | 'blurhash' | 'title' | 'description'
  >
}

const TimelineCard = React.memo(({ card }: Props) => {
  const { colors } = useTheme()
  const navigation = useNavigation()

  return (
    <Pressable
      accessible
      accessibilityRole='link'
      style={[styles.card, { borderColor: colors.border }]}
      onPress={async () => {
        analytics('timeline_shared_card_press')
        await openLink(card.url, navigation)
      }}
      testID='base'
    >
      {card.image ? (
        <GracefullyImage
          uri={{ original: card.image }}
          blurhash={card.blurhash}
          style={styles.left}
          imageStyle={styles.image}
        />
      ) : null}
      <View style={styles.right}>
        <Text
          numberOfLines={2}
          style={[styles.rightTitle, { color: colors.primaryDefault }]}
          testID='title'
        >
          {card.title}
        </Text>
        {card.description ? (
          <Text
            numberOfLines={1}
            style={[styles.rightDescription, { color: colors.primaryDefault }]}
            testID='description'
          >
            {card.description}
          </Text>
        ) : null}
        <Text
          numberOfLines={1}
          style={[styles.rightLink, { color: colors.secondary }]}
        >
          {card.url}
        </Text>
      </View>
    </Pressable>
  )
})

const styles = StyleSheet.create({
  card: {
    flex: 1,
    flexDirection: 'row',
    height: StyleConstants.Font.LineHeight.M * 5,
    marginTop: StyleConstants.Spacing.M,
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 6,
    overflow: 'hidden'
  },
  left: {
    flexBasis: StyleConstants.Font.LineHeight.M * 5
  },
  image: {
    borderTopLeftRadius: 6,
    borderBottomLeftRadius: 6
  },
  right: {
    flex: 1,
    padding: StyleConstants.Spacing.S
  },
  rightTitle: {
    ...StyleConstants.FontStyle.S,
    marginBottom: StyleConstants.Spacing.XS,
    fontWeight: StyleConstants.Font.Weight.Bold
  },
  rightDescription: {
    ...StyleConstants.FontStyle.S,
    marginBottom: StyleConstants.Spacing.XS
  },
  rightLink: {
    ...StyleConstants.FontStyle.S
  }
})

export default TimelineCard
