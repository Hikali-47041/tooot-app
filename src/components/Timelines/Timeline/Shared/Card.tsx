import React, { useEffect, useMemo, useState } from 'react'
import { Image, Pressable, StyleSheet, Text, View } from 'react-native'
import openLink from '@components/openLink'
import { StyleConstants } from '@utils/styles/constants'
import { useTheme } from '@utils/styles/ThemeManager'
import { Surface } from 'gl-react-expo'
import { Blurhash } from 'gl-react-blurhash'

export interface Props {
  card: Mastodon.Card
}

type CancelPromise = ((reason?: Error) => void) | undefined
type ImageSize = { width: number; height: number }
interface ImageSizeOperation {
  start: () => Promise<ImageSize>
  cancel: CancelPromise
}
const getImageSize = (uri: string): ImageSizeOperation => {
  let cancel: CancelPromise
  const start = (): Promise<ImageSize> =>
    new Promise<{ width: number; height: number }>((resolve, reject) => {
      cancel = reject
      Image.getSize(
        uri,
        (width, height) => {
          cancel = undefined
          resolve({ width, height })
        },
        error => {
          reject(error)
        }
      )
    })

  return { start, cancel }
}

const TimelineCard: React.FC<Props> = ({ card }) => {
  const { theme } = useTheme()

  const [imageLoaded, setImageLoaded] = useState(false)
  useEffect(() => {
    if (card.image) {
      Image.getSize(card.image, () => setImageLoaded(true))
    }
  }, [])
  useEffect(() => {
    let cancel: CancelPromise
    const sideEffect = async (): Promise<void> => {
      try {
        const operation = getImageSize(card.image)
        cancel = operation.cancel
        await operation.start()
      } catch (error) {
        if (__DEV__) console.warn(error)
      }
    }
    if (card.image) {
      sideEffect()
    }
    return () => {
      if (cancel) {
        cancel()
      }
    }
  })
  const cardVisual = useMemo(() => {
    if (imageLoaded) {
      return <Image source={{ uri: card.image }} style={styles.image} />
    } else {
      return card.blurhash ? (
        <Surface style={styles.image}>
          <Blurhash hash={card.blurhash} />
        </Surface>
      ) : null
    }
  }, [imageLoaded])

  return (
    <Pressable
      style={[styles.card, { borderColor: theme.border }]}
      onPress={async () => await openLink(card.url)}
      testID='base'
    >
      {card.image && (
        <View style={styles.left} testID='image'>
          {cardVisual}
        </View>
      )}
      <View style={styles.right}>
        <Text
          numberOfLines={2}
          style={[styles.rightTitle, { color: theme.primary }]}
          testID='title'
        >
          {card.title}
        </Text>
        {card.description ? (
          <Text
            numberOfLines={1}
            style={[styles.rightDescription, { color: theme.primary }]}
            testID='description'
          >
            {card.description}
          </Text>
        ) : null}
        <Text numberOfLines={1} style={{ color: theme.secondary }}>
          {card.url}
        </Text>
      </View>
    </Pressable>
  )
}

const styles = StyleSheet.create({
  card: {
    flex: 1,
    flexDirection: 'row',
    height: StyleConstants.Avatar.L,
    marginTop: StyleConstants.Spacing.M,
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 6
  },
  left: {
    width: StyleConstants.Avatar.L
  },
  image: {
    width: '100%',
    height: '100%',
    borderTopLeftRadius: 6,
    borderBottomLeftRadius: 6
  },
  right: {
    flex: 1,
    padding: StyleConstants.Spacing.S
  },
  rightTitle: {
    marginBottom: StyleConstants.Spacing.XS,
    fontWeight: StyleConstants.Font.Weight.Bold
  },
  rightDescription: {
    marginBottom: StyleConstants.Spacing.XS
  }
})

export default React.memo(TimelineCard, () => true)
