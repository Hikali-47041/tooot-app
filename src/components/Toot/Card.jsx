import React from 'react'
import PropTypes from 'prop-types'
import { Image, Pressable, StyleSheet, Text, View } from 'react-native'
import { useNavigation } from '@react-navigation/native'

export default function Card ({ card }) {
  const navigation = useNavigation()
  return (
    card && (
      <Pressable
        style={styles.card}
        onPress={() => {
          navigation.navigate('Webview', {
            uri: card.url
          })
        }}
      >
        {card.image && (
          <View style={styles.left}>
            <Image source={{ uri: card.image }} style={styles.image} />
          </View>
        )}
        <View style={styles.right}>
          <Text numberOfLines={1}>{card.title}</Text>
          {card.description ? (
            <Text numberOfLines={2}>{card.description}</Text>
          ) : (
            <></>
          )}
          <Text numberOfLines={1}>{card.url}</Text>
        </View>
      </Pressable>
    )
  )
}

const styles = StyleSheet.create({
  card: {
    flex: 1,
    flexDirection: 'row',
    height: 70,
    marginTop: 12
  },
  left: {
    width: 70
  },
  image: {
    width: '100%',
    height: '100%'
  },
  right: {
    flex: 1
  }
})

Card.propTypes = {
  card: PropTypes.exact({
    url: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    type: PropTypes.oneOf(['link', 'photo', 'video']),
    author_name: PropTypes.string,
    author_url: PropTypes.string,
    provider_name: PropTypes.string,
    provider_url: PropTypes.string,
    html: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    image: PropTypes.string,
    embed_url: PropTypes.string,
    blurhash: PropTypes.string
  }).isRequired
}
