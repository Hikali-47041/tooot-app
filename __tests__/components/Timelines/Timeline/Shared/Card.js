import React from 'react'
import {
  toBeDisabled,
  toContainElement,
  toHaveStyle,
  toHaveTextContent
} from '@testing-library/jest-native'
import { cleanup, render } from '@testing-library/react-native/pure'

import Card from '@components/Timelines/Timeline/Shared/Card'

expect.extend({
  toBeDisabled,
  toContainElement,
  toHaveStyle,
  toHaveTextContent
})

describe('Testing component timeline card', () => {
  afterEach(cleanup)

  it('with text only', () => {
    const { getByTestId, queryByTestId, toJSON } = render(
      <Card
        card={{
          url: 'http://example.com',
          title: 'Title'
        }}
      />
    )

    expect(queryByTestId('image')).toBeNull()
    expect(getByTestId('base')).toContainElement(getByTestId('title'))
    expect(queryByTestId('description')).toBeNull()

    expect(getByTestId('title')).toHaveTextContent('Title')
    expect(toJSON()).toMatchSnapshot()
  })

  it('with text and description', () => {
    const { getByTestId, queryByTestId, toJSON } = render(
      <Card
        card={{
          url: 'http://example.com',
          title: 'Title',
          description: 'Description'
        }}
      />
    )

    expect(queryByTestId('image')).toBeNull()
    expect(getByTestId('base')).toContainElement(getByTestId('title'))
    expect(getByTestId('base')).toContainElement(getByTestId('description'))

    expect(getByTestId('title')).toHaveTextContent('Title')
    expect(getByTestId('description')).toHaveTextContent('Description')
    expect(toJSON()).toMatchSnapshot()
  })
})
