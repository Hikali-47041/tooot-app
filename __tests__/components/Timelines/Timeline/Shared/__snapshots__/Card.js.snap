// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing component timeline card with text and description 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 6,
        "borderWidth": 0.5,
        "flex": 1,
        "flexDirection": "row",
        "height": 104,
        "marginTop": 16,
      },
      Object {
        "borderColor": "rgba(18, 18, 18, 0.3)",
      },
    ]
  }
  testID="base"
>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 8,
      }
    }
  >
    <Text
      numberOfLines={2}
      style={
        Array [
          Object {
            "fontWeight": "600",
            "marginBottom": 4,
          },
          Object {
            "color": "rgb(18, 18, 18)",
          },
        ]
      }
      testID="title"
    >
      Title
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "marginBottom": 4,
          },
          Object {
            "color": "rgb(18, 18, 18)",
          },
        ]
      }
      testID="description"
    >
      Description
    </Text>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "rgb(135, 135, 135)",
        }
      }
    >
      http://example.com
    </Text>
  </View>
</View>
`;

exports[`Testing component timeline card with text only 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 6,
        "borderWidth": 0.5,
        "flex": 1,
        "flexDirection": "row",
        "height": 104,
        "marginTop": 16,
      },
      Object {
        "borderColor": "rgba(18, 18, 18, 0.3)",
      },
    ]
  }
  testID="base"
>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 8,
      }
    }
  >
    <Text
      numberOfLines={2}
      style={
        Array [
          Object {
            "fontWeight": "600",
            "marginBottom": 4,
          },
          Object {
            "color": "rgb(18, 18, 18)",
          },
        ]
      }
      testID="title"
    >
      Title
    </Text>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "rgb(135, 135, 135)",
        }
      }
    >
      http://example.com
    </Text>
  </View>
</View>
`;
